practice/flutter/flutter/logic.py

```diff
@@ -0,0 +1,167 @@
+    if not matches:
+        raise LookupError('No matches')
+    else:
+        return matches
+
+
+def login_user(request, username, password):
+    """Logs the user in.
+
+    Returns a LookupError if the username does not exist.
+
+    Returns a ValueError if the password was incorrect.
+
+    >>> user = User.objects.create_user('john', 'lennon@thebeatles.com', 'johnpassword')
+    >>> c = client.Client()
+    >>> c.login(username='john', password='johnpassword')
+    True
+    >>> login_user('request', 'john', 'floorb')
+    Traceback (most recent call last):
+    ...
+    ValueError: Incorrect password
+    >>> login_user('request', 'fee', 'foo')
+    Traceback (most recent call last):
+    ...
+    LookupError: That user does not exist.
+    """
+    user = authenticate(username=username, password=password)
+    if user is not None:
+        login(request, user)
+    try:
+        User.objects.get(username=username)
+    except User.DoesNotExist:
+        raise LookupError('That user does not exist.')
+    else:
+        raise ValueError('Incorrect password')
+
+
+def get_username_display(username):
+    """Takes in the user object tied to the request.
+
+    If the user is logged in, returns a string which notifies user of what they are logged in as.
+
+    If the user is not logged in, returns a string which notifies user that they are not logged in.
+
+    >>> get_username_display('goofus')
+    'You are currently logged in as goofus'
+    >>> get_username_display('')
+    'You are currently not logged in. Please login to add a Flutt.'
+    """
+    if username != '':
+        return "You are currently logged in as " + username
+    else:
+        return "You are currently not logged in. Please login to add a Flutt."
+
+
+def get_user_id(username):
+    """Takes in a username, returns the user id for the user who has that username.
+
+    If there is no user with that username, raises a LookupError.
+    >>> user = User.objects.create_user('john', 'lennon@thebeatles.com', 'johnpassword')
+    >>> user.save()
+    >>> get_user_id('john')
+    1
+    >>> get_user_id('food')
+    Traceback (most recent call last):
+    ...
+    LookupError: That user does not exist
+
+    """
+    try:
+        user = User.objects.get(username=username)
+        return user.id
+    except User.DoesNotExist:
+        raise LookupError('That user does not exist')
+
+
+def search_by_user_id(user_id):
+    """Takes an author id, searches for all Flutts which have that author id.
+
+    If there are no Flutts by that user, raises a LookupError.
+
+    >>> models.Flutt(author='Fred', body='2', timestamp=datetime.datetime(2016, 9, 2, 10, 28, 34), author_id=0).save()
+    >>> search_by_user_id(0)
+    <QuerySet [Flutt(author='Fred', body='2', timestamp=2016-09-02 10:28:34+00:00, author_id=0)]>
+    """
+    matches = models.Flutt.objects.filter(author_id=user_id)
+    if not matches:
+        raise LookupError('No Flutts for this user.')
```

Comment:
This is another one of those instances in which you're requiring out-of-band handling of a normal bit of data. If there are no matches, matches will be empty. That's fine! By introducing throwing this exception, all of your other code has to handle this special case. It's already written to handle some number of Flutts.

If you just care about having special text in the template, you can check the length of something and use an `{% if %}` tag.